### PR TITLE
planner: fix correlated aggregates which should be evaluated in outer query (#21431)

### LINKED ISSUE #17748: P1-[4.0 bug hunting]-[Aggregate Fcuntion]-`group_concat` incompatible with MySQL
Bug Hunter issue tidb-challenge-program/bug-hunting-issue#78

---

## Bug Report

### 1. What did you do?

````
drop table if exists t1, t2;

create table t1 (a int, b int);
create table t2 (m int, n int);
insert into t1 values (2,2), (2,2), (3,3), (3,3), (3,3), (4,4);
insert into t2 values (1,11), (2,22), (3,32), (4,44), (4,44);
set @@sql_mode='TRADITIONAL';

select count(*) c, a,
 ( select group_concat(count(a)) from t2 where m = a )
 from t1 group by a;
````

### 2. What did you expect to see?
````
mysql> select count(*) c, a,
    ->  ( select group_concat(count(a)) from t2 where m = a )
    ->  from t1 group by a;
+---+------+-------------------------------------------------------+
| c | a    | ( select group_concat(count(a)) from t2 where m = a ) |
+---+------+-------------------------------------------------------+
| 2 |    2 | 2                                                     |
| 3 |    3 | 3                                                     |
| 1 |    4 | 1,1                                                   |
+---+------+-------------------------------------------------------+
3 rows in set (0.00 sec)
````

### 3. What did you see instead?

````
mysql> select count(*) c, a,
    ->  ( select group_concat(count(a)) from t2 where m = a )
    ->  from t1 group by a;
ERROR 1111 (HY000): Invalid use of group function
````

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

````
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                           |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: None
Edition: None
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
````

Reproducible on master branch:
````
commit 2efab88a59499ea3934c8c1eba7b695cf24b9d29 (HEAD -> master, origin/master, origin/HEAD)
Author: Wenxuan <hi@breeswish.org>
Date:   Fri May 15 21:53:32 2020 +0800
````

### LINKED ISSUE #18350: wrong result when evaluating correlated aggregate functions
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
``` sql
CREATE TABLE tab(i INT); 
CREATE TABLE tab2(j INT); 

insert into tab values(1),(2),(3);
insert into tab2 values(1),(2),(3),(15);

SELECT m.i, 
       (SELECT COUNT(n.j) 
           FROM tab2 WHERE j=15) AS o 
    FROM tab m, tab2 n GROUP BY 1 order by m.i;
```

### 2. What did you expect to see? (Required)
``` sql
mysql> SELECT m.i,
         (SELECT COUNT(n.j)
             FROM tab2 WHERE j=15) AS o
      FROM tab m, tab2 n GROUP BY 1 order by m.i;
+------+------+
| i    | o    |
+------+------+
|    1 |    4 |
|    2 |    4 |
|    3 |    4 |
+------+------+
3 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)
``` sql
tidb> SELECT m.i,
         (SELECT COUNT(n.j)
             FROM tab2 WHERE j=15) AS o
      FROM tab m, tab2 n GROUP BY 1 order by m.i;
+------+---+
| i    | o |
+------+---+
|    1 | 1 |
|    2 | 1 |
|    3 | 1 |
+------+---+
3 rows in set (0.00 sec)
```

### 4. Affected version (Required)

all the versions

### 5. Root Cause Analysis

TiDB does not handle correlated aggregate in subqueries correctly. I do not find any references from MySQL official manual, but [this doc](https://www.ibm.com/support/knowledgecenter/SSGU8G_12.1.0/com.ibm.sqls.doc/ids_sqs_2115.htm) can be referred.
## SIG slack channel

 [#sig-planner](https://slack.tidb.io/invite?team=tidb-community&channel=sig-planner&ref=high-performance)

## Score

- 900

## Mentor

* @lzmhhh123

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #21431 to release-4.0

---

### What problem does this PR solve?

Issue Number: close #18350, close #17748 

Problem Summary:

- Some correlated aggregate (belonging to outer query) should be evaluated in outer query instead of in subquery.

``` mysql
# These aggregates should be evaluated in outer query
select (select count(a)) from t;
select (select count(n.a) from t limit 1) from t n;
select (select 1 from t where count(n.a) > 1 limit 1) from t n;
select (select 1 from t order by count(n.a) limit 1) from t n;
select (select 1 from t having count(n.a) > 1 limit 1) from t n;
select (select cnt from (select count(a) as cnt) n) from t;

# These aggregates should be evaluated in sub-query
select (select count(a + n.a) from t) from t n;
select (select count(a) from t) from t n;
```

- Cannot handle nested aggregates.

``` mysql
select (select sum(count(a))) from t; # report "invalid group function use"
```

### What is changed and how it works?

What's Changed:

- Add a procedure `resolveCorrelatedAggregates` before `buildAggregation` to collect correlated aggregates from sub-queries.
- Add cache for `ResultSetNode` to avoid rebuilding plans.
- Add `correlatedAggMap` for sub-queries to build correct plans.

How it Works:

1. For example, we have

``` mysql
create table t (a int, b int);
```

2. Start from query:

``` mysql
select (select count(n.a) from t limit 1) from t n;
```

3. Inside `b.resolveCorrelatedAggregates`, `count(n.a)` will be recognized as correlated aggregate since column `n.a` is from outer schema (from table `n` instead of `t`). So `count(n.a)` is collected and returned to outer `buildSelect`, now the query is like:

``` mysql
select (select count(n.a) from t limit 1), count(n.a) as `sel_subq_agg_1` from t n;
```

4. Inside `b.buildAggregation`, a correlated column for `sel_subq_agg_1` will be created and added to `b.correlatedAggMap`. Then `b.buildProjection` will build plan for sub-query `(select count(n.a) from t) limit 1`.

5. The sub-query will recognize `count(n.a)` as correlated aggregate and skip it the process of building aggregate plan. Then inside `b.buildProjection` of the sub-query, `count(n.a)` will be rewritten as a correlated column to `sel_subq_agg_1`, now the query is like:

``` mysql
select (select CorrelatedColumn{`sel_subq_agg_1`} from t limit 1), count(n.a) as `sel_subq_agg_1` from t n;
```

6. Finally we get the correct query plan.

### Related changes

- Need to cherry-pick to the release branch 4.0

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- Fix correlated aggregates which should be evaluated in outer query instead of in subqueries.
- Support nested aggregate in subqueries (correlated aggregates inside normal aggregate).

### CODE REVIEW

File: planner/core/logical_plan_test.go [hunk 1]
```diff
@@ -1670,6 +1670,53 @@ func (s *testPlanSuite) TestSimplyOuterJoinWithOnlyOuterExpr(c *C) {
 	c.Assert(join.JoinType, Equals, RightOuterJoin)
 }
 
+func (s *testPlanSuite) TestResolvingCorrelatedAggregate(c *C) {
```
Comment (qw4990) [inline planner/core/logical_plan_test.go:1673]: Should we also modify `checkDataSourceCols` like https://github.com/pingcap/tidb/pull/21431/files?file-filters%5B%5D=.go#diff-037390a673963744ffeb2623d0f17bc5205f52510dda89f3522d44309621f7fbR621?

**Review by qw4990 — COMMENTED** [review]
Rest LGTM

**Review by qw4990 — COMMENTED** [review]
LGTM

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**